controllers/user/login.js

```diff
@@ -44,33 +58,63 @@ router
         });
         renderForm(req, res);
     })
-    .post((req, res, next) => {
+    .post(async (req, res, next) => {
         logger.info('Login attempted');
-        passport.authenticate('local', function(err, user) {
-            if (err) {
-                logger.error('Authentication failed', err);
-                next(err);
-            } else if (user) {
-                req.logIn(user, function(loginErr) {
-                    if (loginErr) {
-                        logger.error('Login failed', loginErr);
-                        next(loginErr);
-                    } else {
-                        logger.info('Login succeeded');
-                        redirectUrlWithFallback(req, res, '/user');
+
+        // Key all rate-limited login attempts by the user's IP and the attempted username
+        const uniqueKey = `${req.ip}_${req.body.username}`;
+
+        const LoginRateLimiter = await new RateLimiter(
+            availableRateLimiters.failByUsername,
+            uniqueKey
+        ).init();
+
+        if (LoginRateLimiter.isRateLimited()) {
+            // User is rate limited
+            return renderRateLimitError(req, res);
+        } else {
+            passport.authenticate('local', async function(err, user) {
+                if (err) {
+                    logger.error('Authentication failed', err);
+                    next(err);
+                } else if (user) {
+                    req.logIn(user, async function(loginErr) {
+                        if (loginErr) {
+                            logger.error('Login failed', loginErr);
+                            next(loginErr);
+                        } else {
+                            logger.info('Login succeeded');
+                            // Reset rate limit on successful authorisation
+                            if (LoginRateLimiter.hasConsumedPoints()) {
+                                await LoginRateLimiter.clearRateLimit();
+                            }
+                            redirectUrlWithFallback(req, res, '/user');
+                        }
+                    });
+                } else {
+                    /**
+                     * User is invalid
+                     * Show a generic error message here to avoid exposing account state
+                     */
+                    logger.warn('Login failed: invalid credentials');
+                    try {
+                        await LoginRateLimiter.consumeRateLimit();
+                        return renderForm(req, res, req.body, [
+                            {
+                                msg: `Your username and password combination is invalid`
+                            }
+                        ]);
+                    } catch (rlRejected) {
```

Comment:
Mind renaming to something like `rateLimitRejection`, took me a bit to grok what `rlRejected ` was. Also is there a clearer way of determining the type of error than checking `instanceof`?

Reply:
This was copied straight from the docs for this module so I guess it's their way of determining that this is specifically an error with the module (eg. I triggered this by deleting one of its table rows midway through its validation process) rather than anything else. Will rename though!

Reply:
Ahh, looking at it again, they have this example:

```
} catch (rlRejected) {
        if (rlRejected instanceof Error) {
          throw rlRejected;
        } else {
          res.set('Retry-After', String(Math.round(rlRejected.msBeforeNext / 1000)) || 1);
          res.status(429).send('Too Many Requests');
        }
      }
```

So sometimes this is an actual error (eg. something broke unexpectedly), other times it contains details of the user's current rate limit.